File: packages/shared-types/src/legal.ts
/** Legal & compliance types shared between frontend and backend. */

// ─── Consent Types ────────────────────────────────────────────

export type ConsentType =
    | "home_spital"
    | "iv_antibiotics"
    | "telemedizin"
    | "ndsg"
    | "epdg"
    | "thromboprophylaxe";

export type ConsentStatus = "pending" | "granted" | "refused" | "revoked";

export const CONSENT_TYPE_LABELS: Record<ConsentType, string> = {
    home_spital: "Home-Spital Behandlung",
    iv_antibiotics: "IV-Antibiose zu Hause",
    telemedizin: "Telemedizin-Konsultation",
    ndsg: "Datenschutz (nDSG)",
    epdg: "Elektronisches Patientendossier (EPDG)",
    thromboprophylaxe: "Thromboseprophylaxe",
};

export const CONSENT_STATUS_LABELS: Record<ConsentStatus, string> = {
    pending: "Ausstehend",
    granted: "Erteilt",
    refused: "Verweigert",
    revoked: "Widerrufen",
};

export interface Consent {
    id: string;
    patient_id: string;
    consent_type: ConsentType;
    status: ConsentStatus;
    granted_at?: string;
    granted_by?: string;
    valid_from?: string;
    valid_until?: string;
    revoked_at?: string;
    revoked_reason?: string;
    witness_name?: string;
    notes?: string;
    created_at: string;
    updated_at: string;
}

export interface ConsentCreate {
    patient_id: string;
    consent_type: ConsentType;
    status?: ConsentStatus;
    granted_by?: string;
    valid_from?: string;
    valid_until?: string;
    witness_name?: string;
    notes?: string;
}

export interface ConsentUpdate {
    status?: ConsentStatus;
    granted_by?: string;
    valid_from?: string;
    valid_until?: string;
    witness_name?: string;
    notes?: string;
}

export interface ConsentMeta {
    consent_types: Record<string, string>;
    consent_statuses: Record<string, string>;
}

// ─── Advance Directive Types ──────────────────────────────────

export type DirectiveType = "patientenverfuegung" | "vorsorgeauftrag";
export type ReaStatus = "FULL" | "DNR";

export const DIRECTIVE_TYPE_LABELS: Record<DirectiveType, string> = {
    patientenverfuegung: "Patientenverfügung",
    vorsorgeauftrag: "Vorsorgeauftrag",
};

export interface AdvanceDirective {
    id: string;
    patient_id: string;
    directive_type: DirectiveType;
    rea_status: ReaStatus;
    intensive_care: boolean;
    mechanical_ventilation: boolean;
    dialysis: boolean;
    artificial_nutrition: boolean;
    trusted_person_name?: string;
    trusted_person_phone?: string;
    trusted_person_relation?: string;
    trusted_person_contact_id?: string;
    document_date?: string;
    storage_location?: string;
    is_valid: boolean;
    notes?: string;
    created_at: string;
    updated_at: string;
}

export interface DirectiveCreate {
    patient_id: string;
    directive_type: DirectiveType;
    rea_status: ReaStatus;
    intensive_care?: boolean;
    mechanical_ventilation?: boolean;
    dialysis?: boolean;
    artificial_nutrition?: boolean;
    trusted_person_name?: string;
    trusted_person_phone?: string;
    trusted_person_relation?: string;
    trusted_person_contact_id?: string;
    document_date?: string;
    storage_location?: string;
    notes?: string;
}

export interface DirectiveUpdate {
    directive_type?: DirectiveType;
    rea_status?: ReaStatus;
    intensive_care?: boolean;
    mechanical_ventilation?: boolean;
    dialysis?: boolean;
    artificial_nutrition?: boolean;
    trusted_person_name?: string;
    trusted_person_phone?: string;
    trusted_person_relation?: string;
    trusted_person_contact_id?: string;
    document_date?: string;
    storage_location?: string;
    is_valid?: boolean;
    notes?: string;
}

export interface DirectiveMeta {
    directive_types: Record<string, string>;
    rea_statuses: Record<string, string>;
}

// ─── Patient Wishes Types (ZGB 378) ──────────────────────────

export interface PatientWishes {
    id: string;
    patient_id: string;
    quality_of_life?: string;
    autonomy_preferences?: string;
    pain_management?: string;
    decision_maker?: string;
    decision_maker_contact_id?: string;
    sleep_preferences?: string;
    nutrition_preferences?: string;
    family_wishes?: string;
    pet_info?: string;
    spiritual_needs?: string;
    other_wishes?: string;
    recorded_by?: string;
    recorded_at: string;
    updated_at: string;
}

export interface WishesUpsert {
    quality_of_life?: string;
    autonomy_preferences?: string;
    pain_management?: string;
    decision_maker?: string;
    decision_maker_contact_id?: string;
    sleep_preferences?: string;
    nutrition_preferences?: string;
    family_wishes?: string;
    pet_info?: string;
    spiritual_needs?: string;
    other_wishes?: string;
}

// ─── Palliative Care Types ────────────────────────────────────

export interface PalliativeCare {
    id: string;
    patient_id: string;
    is_active: boolean;
    activated_at?: string;
    activated_by?: string;
    reserve_morphin?: string;
    reserve_midazolam?: string;
    reserve_haloperidol?: string;
    reserve_scopolamin?: string;
    reserve_other?: string;
    palliative_service_name?: string;
    palliative_service_phone?: string;
    palliative_service_email?: string;
    goals_of_care?: string;
    notes?: string;
    updated_at: string;
}

export interface PalliativeUpsert {
    is_active?: boolean;
    reserve_morphin?: string;
    reserve_midazolam?: string;
    reserve_haloperidol?: string;
    reserve_scopolamin?: string;
    reserve_other?: string;
    palliative_service_name?: string;
    palliative_service_phone?: string;
    palliative_service_email?: string;
    goals_of_care?: string;
    notes?: string;
}

// ─── Death Notification Types ─────────────────────────────────

export interface DeathNotification {
    id: string;
    patient_id: string;
    contact_name: string;
    contact_phone?: string;
    contact_email?: string;
    contact_role?: string;
    emergency_contact_id?: string;
    priority: number;
    instructions?: string;
    created_at: string;
}

export interface DeathNotificationCreate {
    patient_id: string;
    contact_name: string;
    contact_phone?: string;
    contact_email?: string;
    contact_role?: string;
    emergency_contact_id?: string;
    priority?: number;
    instructions?: string;
}

export interface DeathNotificationUpdate {
    contact_name?: string;
    contact_phone?: string;
    contact_email?: string;
    contact_role?: string;
    priority?: number;
    instructions?: string;
}

File: backend/src/domain/schemas/legal.py
"""Legal Pydantic schemas — Consent, AdvanceDirective, PatientWishes, PalliativeCare, DeathNotification."""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


# ════════════════════════════════════════════════════════════════
# Consent (Einwilligung)
# ════════════════════════════════════════════════════════════════

CONSENT_TYPES = (
    "home_spital", "iv_antibiotics", "telemedizin",
    "ndsg", "epdg", "thromboprophylaxe",
)

CONSENT_TYPE_LABELS: dict[str, str] = {
    "home_spital": "Home-Spital-Behandlung",
    "iv_antibiotics": "i.v.-Antibiotikatherapie",
    "telemedizin": "Telemedizin / Teleconsult",
    "ndsg": "Datenbearbeitung (nDSG)",
    "epdg": "Elektronisches Patientendossier (EPDG)",
    "thromboprophylaxe": "Thromboseprophylaxe",
}

CONSENT_STATUS_LABELS: dict[str, str] = {
    "pending": "Ausstehend",
    "granted": "Erteilt",
    "refused": "Verweigert",
    "revoked": "Widerrufen",
}


class ConsentCreate(BaseModel):
    patient_id: uuid.UUID
    consent_type: str = Field(..., pattern=r"^(home_spital|iv_antibiotics|telemedizin|ndsg|epdg|thromboprophylaxe)$")
    status: str = Field("pending", pattern=r"^(pending|granted|refused)$")
    granted_at: datetime | None = None
    granted_by: str | None = None
    valid_from: date | None = None
    valid_until: date | None = None
    witness_name: str | None = None
    notes: str | None = None


class ConsentUpdate(BaseModel):
    status: str | None = Field(None, pattern=r"^(pending|granted|refused|revoked)$")
    granted_at: datetime | None = None
    granted_by: str | None = None
    valid_from: date | None = None
    valid_until: date | None = None
    revoked_at: datetime | None = None
    revoked_reason: str | None = None
    witness_name: str | None = None
    notes: str | None = None


class ConsentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    patient_id: uuid.UUID
    consent_type: str
    status: str
    granted_at: datetime | None
    granted_by: str | None
    valid_from: date | None
    valid_until: date | None
    revoked_at: datetime | None
    revoked_reason: str | None
    witness_name: str | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


class PaginatedConsents(BaseModel):
    items: list[ConsentResponse]
    total: int


# ════════════════════════════════════════════════════════════════
# AdvanceDirective (Patientenverfügung)
# ════════════════════════════════════════════════════════════════

DIRECTIVE_TYPE_LABELS: dict[str, str] = {
    "patientenverfuegung": "Patientenverfügung (FMH)",
    "vorsorgeauftrag": "Vorsorgeauftrag (ZGB Art. 360)",
}


class DirectiveCreate(BaseModel):
    patient_id: uuid.UUID
    directive_type: str = Field(..., pattern=r"^(patientenverfuegung|vorsorgeauftrag)$")
    rea_status: str = Field("FULL", pattern=r"^(FULL|DNR)$")
    intensive_care: bool = True
    mechanical_ventilation: bool = True
    dialysis: bool = True
    artificial_nutrition: bool = True
    trusted_person_name: str | None = None
    trusted_person_phone: str | None = None
    trusted_person_relation: str | None = None
    trusted_person_contact_id: uuid.UUID | None = None
    document_date: date | None = None
    storage_location: str | None = None
    notes: str | None = None


class DirectiveUpdate(BaseModel):
    rea_status: str | None = Field(None, pattern=r"^(FULL|DNR)$")
    intensive_care: bool | None = None
    mechanical_ventilation: bool | None = None
    dialysis: bool | None = None
    artificial_nutrition: bool | None = None
    trusted_person_name: str | None = None
    trusted_person_phone: str | None = None
    trusted_person_relation: str | None = None
    trusted_person_contact_id: uuid.UUID | None = None
    document_date: date | None = None
    storage_location: str | None = None
    is_valid: bool | None = None
    notes: str | None = None


class DirectiveResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    patient_id: uuid.UUID
    directive_type: str
    rea_status: str
    intensive_care: bool
    mechanical_ventilation: bool
    dialysis: bool
    artificial_nutrition: bool
    trusted_person_name: str | None
    trusted_person_phone: str | None
    trusted_person_relation: str | None
    trusted_person_contact_id: uuid.UUID | None
    document_date: date | None
    storage_location: str | None
    is_valid: bool
    notes: str | None
    created_at: datetime
    updated_at: datetime


# ════════════════════════════════════════════════════════════════
# PatientWishes (Mutmasslicher Wille)
# ════════════════════════════════════════════════════════════════


class WishesUpsert(BaseModel):
    quality_of_life: str | None = None
    autonomy_preferences: str | None = None
    pain_management: str | None = None
    decision_maker: str | None = None
    decision_maker_contact_id: uuid.UUID | None = None
    sleep_preferences: str | None = None
    nutrition_preferences: str | None = None
    family_wishes: str | None = None
    pet_info: str | None = None
    spiritual_needs: str | None = None
    other_wishes: str | None = None


class WishesResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    patient_id: uuid.UUID
    quality_of_life: str | None
    autonomy_preferences: str | None
    pain_management: str | None
    decision_maker: str | None
    decision_maker_contact_id: uuid.UUID | None
    sleep_preferences: str | None
    nutrition_preferences: str | None
    family_wishes: str | None
    pet_info: str | None
    spiritual_needs: str | None
    other_wishes: str | None
    recorded_at: datetime
    updated_at: datetime


# ════════════════════════════════════════════════════════════════
# PalliativeCare
# ════════════════════════════════════════════════════════════════


class PalliativeUpsert(BaseModel):
    is_active: bool | None = None
    reserve_morphin: str | None = None
    reserve_midazolam: str | None = None
    reserve_haloperidol: str | None = None
    reserve_scopolamin: str | None = None
    reserve_other: str | None = None
    palliative_service_name: str | None = None
    palliative_service_phone: str | None = None
    palliative_service_email: str | None = None
    goals_of_care: str | None = None
    notes: str | None = None


class PalliativeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    patient_id: uuid.UUID
    is_active: bool
    activated_at: datetime | None
    activated_by: uuid.UUID | None
    reserve_morphin: str | None
    reserve_midazolam: str | None
    reserve_haloperidol: str | None
    reserve_scopolamin: str | None
    reserve_other: str | None
    palliative_service_name: str | None
    palliative_service_phone: str | None
    palliative_service_email: str | None
    goals_of_care: str | None
    notes: str | None
    updated_at: datetime


# ════════════════════════════════════════════════════════════════
# DeathNotification (Todesfall-Mitteilungen)
# ════════════════════════════════════════════════════════════════


class DeathNotificationCreate(BaseModel):
    patient_id: uuid.UUID
    contact_name: str = Field(..., min_length=1, max_length=200)
    contact_phone: str | None = None
    contact_email: str | None = None
    contact_role: str | None = None
    emergency_contact_id: uuid.UUID | None = None
    priority: int = Field(2, ge=1, le=3)
    instructions: str | None = None


class DeathNotificationUpdate(BaseModel):
    contact_name: str | None = Field(None, max_length=200)
    contact_phone: str | None = None
    contact_email: str | None = None
    contact_role: str | None = None
    emergency_contact_id: uuid.UUID | None = None
    priority: int | None = Field(None, ge=1, le=3)
    instructions: str | None = None


class DeathNotificationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    patient_id: uuid.UUID
    contact_name: str
    contact_phone: str | None
    contact_email: str | None
    contact_role: str | None
    emergency_contact_id: uuid.UUID | None
    priority: int
    instructions: str | None
    created_at: datetime
